lib/stream.js

```diff
@@ -449,18 +450,26 @@ class SyncStream extends Pipeable {
    * @param {Data} data
    */
   write(data) {
+    if (!data) return this.#debug("Invalid data from remote peer: missing data")
+    if (typeof data !== "object") return this.#debug("Invalid data from remote peer: not an object")
+    if (Array.isArray(data)) return this.#debug("Invalid data from remote peer: is an array")
     const { id, phase, payload } = data
+    if (typeof phase !== 'number') return this.#debug("Invalid data from remote peer: phase isn't a number")
+    if (!isMsgId(id)) return this.#debug("Invalid data from remote peer: id is not a valid msg id")
+    if (phase !== 0 && !payload) return this.#debug("Invalid data from remote peer: payload is missing")
 
-    // TODO: validate that each data objects has the exact correct shape
     switch (phase) {
       case 0: {
         return this.#sendLocalHave(id)
       }
       case 1: {
+        if (!isRange(payload)) return this.#debug("Invalid data from remote peer: payload is not a range in phase 1")
         return this.#sendLocalHaveAndWant(id, payload)
       }
       case 2: {
         const { haveRange, wantRange } = payload
+        if (!isRange(haveRange) || !isRange(wantRange)) return this.#debug("Invalid data from remote peer: haveRange or wantRange is not a range in phase 2")
```

Comment:
This very long line is going to get reformatted by prettier, so let's preface these lines with a leading `// prettier-ignore` comment. It's standard practice in these codebases to put `// prettier-ignore` on long error messages and warning like these.

Reply:
did this for all the debugs now